### PR TITLE
display warning about view permissions

### PR DESCRIPTION
**Pull request by StupidCodeFactory**
## CHANGES:

As per the [ticket](https://trello.com/c/G1tdeHxB/547-2-dont-show-enquiry-contact-details-to-teams-not-involved-in-case), now a hint message is going to be displayed when viewing an enquiry under the contact details section in the summary list.
This was a bit tricky to integrate with the existing GDPR restriction on top of this. 

## miscellaneous

1. I had to remove some of the assumption around `User.current` and explicitly pass in the `current_user` as the viewing user to all the methods that do permission like stuff.
2. Cleaned up the `investigations/tabs/_overview.erb`. Some businessy logic was being calculated in the view and were lacking some tests, and it was generally hard to follow what is going on inside between, permissions, collaborating teams and possible actions. Hence I moved these parts to helpers. 
3. removed rendering of details summary list from decorator
4. added a `ApplicationHelper#permissions_hint` that will be useful to display contextual hints. 
 

## Checklist:
- [X] Have you documented your changes in the pull request description?
- [X] Does the change present any security considerations?
- [ ] Has acceptance criteria been tested by a peer?

Screenshot 2020-05-21 at 13 44 18

### CODE REVIEW

**Comment by StupidCodeFactory** [comment]
[(2) Don't show enquiry contact details to teams not involved in case](https://trello.com/c/G1tdeHxB/547-2-dont-show-enquiry-contact-details-to-teams-not-involved-in-case)

File: psd-web/spec/factories/complainants.rb [hunk 1]
```diff
@@ -5,7 +5,6 @@
     other_details { Faker::Lorem.paragraph }
     phone_number { Faker::PhoneNumber.phone_number }
     complainant_type { Complainant::TYPES.keys.sample }
-    investigation
```
Comment (StupidCodeFactory) [inline psd-web/spec/factories/complainants.rb:8]: this would unnecessarily create an investigation before  assigning it to another investigation.

File: psd-web/app/helpers/investigations_helper.rb [hunk 2]
```diff
@@ -248,6 +248,34 @@ def suggested_previous_owners
...
+    rows
+  end
+
+  # rubocop:disable Rails/OutputSafety
+  def team_list_html(investigation)
+    return investigation.teams_with_access.first.name if investigation.teams_with_access.length == 1
+
+    tag.ul(class: "govuk-list") do
+      investigation.teams_with_access.map { |team| tag.li(escape_once(team.name)) }.join.html_safe
+    end
```
Comment (frankieroberto) [inline psd-web/app/helpers/investigations_helper.rb:269]: Suggest using `concat` to simplify this, and avoid having to explicitly escape and mark as `html_safe`: 

```suggestion
    tag.ul(class: "govuk-list") do
      investigation.teams_with_access do |team|
        concat tag.li(team.name)
      end
    end
```

File: psd-web/app/models/investigation.rb [hunk 3]
```diff
@@ -146,10 +146,10 @@ def reported_reason
     @reported_reason ||= ActiveSupport::StringInquirer.new(super)
   end
 
-  def child_should_be_displayed?
+  def child_should_be_displayed?(user)
     # This method is responsible for white-list access for assignee and their team, as described in
     # https://regulatorydelivery.atlassian.net/wiki/spaces/PSD/pages/598933517/Approach+to+case+sensitivity
-    owner.in_same_team_as?(User.current)
+    owner.in_same_team_as?(user)
```
Comment (frankieroberto) [inline psd-web/app/models/investigation.rb:152]: 👏

File: psd-web/app/views/investigations/tabs/_overview.html.erb [hunk 4]
```diff
@@ -24,50 +24,23 @@
     <% if @investigation.complainant %>
       <%= govuk_hr %>
       <h2 class="govuk-heading-m"><%= @investigation.type.demodulize.upcase_first %></h2>
-      <%= @investigation.source_details_summary_list %>
+      <%= render "components/govuk_summary_list", rows: source_details_rows(@investigation, current_user), classes: "govuk-summary-list--no-border" %>
```
Comment (frankieroberto) [inline psd-web/app/views/investigations/tabs/_overview.html.erb:27]: Please change: could you switch to the [helper version](https://github.com/UKGovernmentBEIS/govuk-design-system-rails/blob/master/app/helpers/govuk_design_system/summary_list_helper.rb) of this, rather than the slim partial, as it contains a couple of bug fixes:

```suggestion
      <%= govukSummaryList(rows: source_details_rows(@investigation, current_user), classes: "govuk-summary-list--no-border") %>
```

File: psd-web/app/decorators/investigation/enquiry_decorator.rb [hunk 5]
```diff
@@ -1,14 +1,16 @@
...
       date_received?
     end
 
     def should_display_received_by?
       received_type?
     end
+
+    def hint_for_contact_details
+      h.tag.span(h.permissions_hint("enquiry contact details"), class: "govuk-hint")
+    end
```
Comment (frankieroberto) [inline psd-web/app/decorators/investigation/enquiry_decorator.rb:14]: Please change: move this to a helper (or the view), as I think we've decided against rendering html from decorators.

File: psd-web/app/helpers/investigations_helper.rb [hunk 6]
```diff
@@ -248,6 +248,34 @@ def suggested_previous_owners
...
 
+  def source_details_rows(investigation, viewing_user)
+    rows = [
+      investigation.should_display_date_received? ? { key: { text: "Received date" }, value: { text: investigation.date_received.to_s(:govuk) } } : nil,
+      investigation.should_display_received_by? ? { key: { text: "Received by" }, value: { text: investigation.received_type.upcase_first } } : nil,
+      { key: { text: "Source type" }, value: { text: investigation.complainant.complainant_type } },
+      { key: { text: "Contact details" }, value: { text: investigation.contact_details_for_display(viewing_user) } }
+    ]
+
+    rows.compact!
```
Comment (frankieroberto) [inline psd-web/app/helpers/investigations_helper.rb:259]: Suggest moving this ternary operators into `if` statements to slightly improve readability (and avoid the need to call `rows.compact!`):

```suggestion
    rows = []
    if investigation.should_display_date_received?
      rows << { key: { text: "Received date" }, value: { text: investigation.date_received.to_s(:govuk) }
    end
    if investigation.should_display_received_by?
      rows << { key: { text: "Received by" }, value: { text: investigation.received_type.upcase_first } }
    end
    rows << { key: { text: "Source type" }, value: { text: investigation.complainant.complainant_type } }
    rows << { key: { text: "Contact details" }, value: { text: investigation.contact_details_for_display(viewing_user) } }
```

File: psd-web/spec/factories/investigations.rb [hunk 7]
```diff
@@ -13,6 +13,10 @@
 
     association :owner, factory: :user
 
+    after(:create) do |investigation|
+      create(:"complainant_#{Complainant::TYPES.keys.sample}", investigation: investigation)
+    end
```
Comment (frankieroberto) [inline psd-web/spec/factories/investigations.rb:18]: Does every investigation need a complainant, or can we just added these for the tests that need them? Adding them for every investigation might slow the tests down a fair bit.

File: psd-web/app/helpers/investigations_helper.rb [hunk 8]
```diff
@@ -248,6 +248,34 @@ def suggested_previous_owners
...
+    rows
+  end
+
+  # rubocop:disable Rails/OutputSafety
+  def team_list_html(investigation)
```
Comment (frankieroberto) [inline psd-web/app/helpers/investigations_helper.rb:264]: Not sure we need the `_html` suffix here, as most of the helpers generate HTML.

```suggestion
  def list_of_teams_with_access(investigation)
```

**Comment by edwardhorsford** [comment]
@StupidCodeFactory can you make the pr title a bit more specific? This is about restricting enquiry source contact details.

File: psd-web/app/helpers/application_helper.rb [hunk 9]
```diff
@@ -20,4 +20,8 @@ def markdown(text)
     sanitized_input = sanitize(text, tags: %w[br])
     rc.render(sanitized_input).html_safe # rubocop:disable Rails/OutputSafety
   end
+
+  def permissions_hint(permission = "this")
+    I18n.t("permissions_hint", permission: permission)
+  end
```
[the remark this answers is not in the document]
Comment (edwardhorsford) [reply psd-web/app/helpers/application_helper.rb:26]: What @StupidCodeFactory has called is the `noun` / `target` to be passed to the hint. It is not a permission IMHO.

Examples:

`Only teams addd to the case can view this`
`Only teams addd to the case can view attachments`
`Only teams addd to the case can view enquiry contact details`

File: psd-web/spec/factories/investigations.rb (hunk 7 above)
Comment (edwardhorsford) [reply psd-web/spec/factories/investigations.rb:18]: Complainants are only collected for enquiries I think?

Comment (slorek) [reply psd-web/spec/factories/investigations.rb:18]: Rather than creating this in every instance I added a optional trait in https://github.com/UKGovernmentBEIS/beis-opss-psd/pull/631

**Comment by StupidCodeFactory** [comment]
closing this. This needs complete re-write
